ledger/simulation/resources.go

```diff
@@ -0,0 +1,584 @@
+	if len(a.Apps) >= a.MaxApps {
+		return false
+	}
+
+	if programVersion >= 7 { // appAddressAvailableVersion
+		appAddr := ep.GetApplicationAddress(aid)
+		// nil map lookup is ok
+		_, ok := a.Accounts[appAddr]
+		if ok {
+			// remove the account reference, since it will be made available by this app reference
+			delete(a.Accounts, appAddr)
+			if len(a.Accounts) == 0 {
+				a.Accounts = nil
+			}
+		}
+	}
+
+	if len(a.Accounts)+len(a.Assets)+len(a.Apps)+len(a.Boxes)+a.NumEmptyBoxRefs >= a.MaxTotalRefs {
+		return false
+	}
+	if a.Apps == nil {
+		a.Apps = make(map[basics.AppIndex]struct{})
+	}
+	a.Apps[aid] = struct{}{}
+	return true
+}
+
+func (a *ResourceAssignment) removeAppSlot() bool {
+	if len(a.Apps) >= a.MaxApps || len(a.Accounts)+len(a.Assets)+len(a.Apps)+len(a.Boxes)+a.NumEmptyBoxRefs >= a.MaxTotalRefs {
+		return false
+	}
+	a.MaxApps--
+	a.MaxTotalRefs--
+	if a.MaxAccounts > 0 {
+		a.MaxAccounts--
+	}
+	return true
+}
+
+func (a *ResourceAssignment) hasBox(app basics.AppIndex, name string) bool {
+	// nil map lookup is ok
+	_, ok := a.Boxes[logic.BoxRef{App: app, Name: name}]
+	return ok
+}
+
+func (a *ResourceAssignment) addBox(app basics.AppIndex, name string, readSize, additionalReadBudget, bytesPerBoxRef uint64) bool {
+	usedReadBudget := basics.AddSaturate(a.usedBoxReadBudget(), readSize)
+	// Adding bytesPerBoxRef to account for the new IO budget from adding an additional box ref
+	readBudget := additionalReadBudget + a.boxIOBudget(bytesPerBoxRef) + bytesPerBoxRef
+
+	var emptyRefs int
+	if usedReadBudget > readBudget {
+		// We need to allocate more empty box refs to increase the read budget
+		neededBudget := usedReadBudget - readBudget
+		// Adding (bytesPerBoxRef - 1) to round up
+		emptyRefs = int((neededBudget + bytesPerBoxRef - 1) / bytesPerBoxRef)
+	} else if a.NumEmptyBoxRefs != 0 {
+		surplusBudget := readBudget - usedReadBudget
+		if surplusBudget >= bytesPerBoxRef && readBudget-bytesPerBoxRef >= a.maxWriteBudget {
+			// If we already have enough read budget, remove one empty ref to be replaced by the new
+			// named box ref.
+			emptyRefs = -1
+		}
+	}
+
+	if len(a.Boxes)+a.NumEmptyBoxRefs+emptyRefs >= a.MaxBoxes || len(a.Accounts)+len(a.Assets)+len(a.Apps)+len(a.Boxes)+a.NumEmptyBoxRefs+emptyRefs >= a.MaxTotalRefs {
+		return false
+	}
+	if a.Boxes == nil {
+		a.Boxes = make(map[logic.BoxRef]uint64)
+	}
+	a.Boxes[logic.BoxRef{App: app, Name: name}] = readSize
+	a.NumEmptyBoxRefs += emptyRefs
+	return true
+}
+
+func (a *ResourceAssignment) addEmptyBoxRefsForWriteBudget(usedWriteBudget, additionalWriteBudget, bytesPerBoxRef uint64) bool {
+	writeBudget := additionalWriteBudget + a.boxIOBudget(bytesPerBoxRef)
+	if usedWriteBudget > writeBudget {
+		// Need to allocate more empty box refs
+		overspend := usedWriteBudget - writeBudget
+		extraRefs := int((overspend + bytesPerBoxRef - 1) / bytesPerBoxRef) // adding (bytesPerBoxRef - 1) to round up
```

Comment:
I supposed you could use `logic.divCeil` here (and above for reads), but you'd have to make it generic for handling `uint64`.

Have you though through overflow as you convert "down" to int? I guess we're probably safe because boxes can only be 32k, and you'll error out once a txn creates more than 8 per txn.

Reply:
I think if that is appealing, you could move it over to `basics`

Reply:
Good point about `divCeil`, I wasn't aware of it when I wrote this. I'll likely expose it and use it here.

I originally had code which would error/panic if the uint64 couldn't be converted to int, but it seemed like overkill because I wasn't really convinced that could even happen at this point. (I also had much more prevalent use of `basics.AddSaturate` and friends, and I abandoned them for similar reasons.)

Reply:
I'm pretty sure I agree on that overflow stuff, but a comment or too explaining the rough math of why somewhere would be welcome. (I think it boils down to the argument above - you only allow so many boxes total because of the limits on num transactions and boxes per txn)